### DISCUSSION TITLE
Parse error on certificate stored in DB

### DISCUSSION THREAD

**Discussion by rabol** [lsnepomuceno/laravel-a1-pdf-sign#84] [Q&A]
Hi

When trying to retrieve a certificate from the MySQL I get this error when calling the parse() method:

```
Process runtime error, reason: "8004387136:error:0DFFF0A8:asn1 encoding routines:
CRYPTO_internal:wrong tag:/AppleInternal/Library/BuildRoots/9e200cfa-7d96-11ed-886f-a23c4f261b56/Library/Caches/com.apple.xbs/Sources/libressl/libressl-3.3/crypto/asn1/tasn_dec.c:1167:. 8004387136:error:0DFFF03A:asn1 encoding routines:
CRYPTO_internal:nested asn1 error:/AppleInternal/Library/BuildRoots/9e200cfa-7d96-11ed-886f-a23c4f261b56/Library/Caches/com.apple.xbs/Sources/libressl/libressl-3.3/crypto/asn1/tasn_dec.c:337:Type=PKCS12. "  
```

Running on a MBP with M1 processor

**Comment by salim-sudo** [comment]
The same error happened to me, with the system windows 10, postgresql 10 and Laravel 10:
LSNepomuceno\LaravelA1PdfSign\Exceptions\ProcessRunTimeException: [0]: Process runtime error, reason: "1828:error:0D0680A8:asn1 encoding routines:asn1_check_tlen:wrong tag:crypto\asn1\tasn_dec.c:1149:. . 1828:error:0D07803A:asn1 encoding routines:asn1_item_embed_d2i:nested asn1 error:crypto\asn1\tasn_dec.c:309:Type=PKCS12. .

**Comment by salim-sudo** [comment]
In my temp folder, the certificate pfx obtained is not encrypted. When I open a certificate, I get the following information:
![tra](https://github.com/user-attachments/assets/45e2e0c2-4b8d-48d7-ac81-73c6f9cc7581)

i have problem in this code

function decryptCertData(
        string $hashKey,
        string $encryptCert,
        string $password,
        bool   $isBase64 = false,
        bool   $usePathEnv = false,
    ): ManageCert
    {
       
        $cert = (new ManageCert)->setHashKey($hashKey);
       
        $uuid = Str::orderedUuid();
        $pfxName = "{$cert->getTempDir()}{$uuid}.pfx";

        $decryptedData = $cert->getEncrypter()->decryptString($encryptCert);
    //  dd((new ManageCert)->setHashKey($hashKey),$encryptCert,$decryptedData,$cert->getEncrypter()->decryptString($password));
        File::put($pfxName, $isBase64 ? base64_decode($decryptedData) : $decryptedData);
  
        return $cert->fromPfx(
            $pfxName,
            $cert->getEncrypter()->decryptString($password),
            $usePathEnv
        );

**Reply by lsnepomuceno** [reply]
Hello @salim-sudo ,

To sign a file with a digital certificate, the decryptString method is not required. This method is only used when encrypting the certificate for storage in a database, for instance.

To use the .pfx file, simply provide the certificate path and password, as shown in the documentation example:
https://laravel-a1-pdf-sign.netlify.app/docs/1.x/working-with-certificate

If you prefer a simpler process, you can just use the helper. In this case, you only need to provide the path to the PDF file, the certificate path, and the password:
https://laravel-a1-pdf-sign.netlify.app/docs/1.x/helpers

**Comment by lsnepomuceno** [comment]
Hello @rabol, @salim-sudo,

This was a real bug in the package, and @salim-sudo's observation — that the `.pfx` written to the temp folder "is not encrypted" — is exactly the diagnosis.

`encryptCertData()` stored `$cert->getCert()->original`, which is the **PEM** bundle produced by `openssl pkcs12 -nodes`. `decryptCertData()` then wrote those PEM bytes to a `.pfx` file and fed it to `openssl pkcs12 -in`, which expects **binary PKCS#12**. Hence:

```
asn1 encoding routines: wrong tag ... Type=PKCS12
```

The pair never round-tripped. Half of the certificate-storage API had never worked, and it went unnoticed because the old test suite only ever *wrote* a certificate and asserted the shape of the return value — it never read one back.

**Fixed in 2.0.** `CertificateVault` stores the PEM bundle and parses it back directly, with no PKCS#12 round-trip and no temporary file:

```php
use LSNepomuceno\LaravelA1PdfSign\Facades\A1PdfSign;

$encrypted = A1PdfSign::encryptCertificate('certificate.pfx', $password);

// store $encrypted->encryptedData, ->encryptedPassword and ->hashKey
// the hash key is required to read the pair back

$certificate = A1PdfSign::decryptCertificate(
    hashKey: $model->hash,
    encryptedCertificate: $model->certificate,
    password: $model->password,
    isBase64: false,   // true only if the column stored a base64 blob
);

A1PdfSign::newSignature()
    ->usingCertificate($certificate)
    ->pdf('document.pdf')
    ->sign();
```

Two practical notes:

- **Store the certificate column as a binary type.** A text column mangles the blob.
- The round trip now has a test that reads back what it wrote, which is what should have caught this originally.

2.0 requires PHP 8.4 and Laravel 13, and the API changed — [UPGRADE.md](https://github.com/lsnepomuceno/laravel-a1-pdf-sign/blob/main/UPGRADE.md) maps every old call to its replacement.

Sorry it stayed open this long.
